api/v1beta1/nodemodulestate_types.go

```diff
@@ -0,0 +1,87 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ModuleNodeSpec struct {
+	Name            string                   `json:"name"`
+	Namespace       string                   `json:"namespace"`
+	ModuleLoader    ModuleLoaderSpec         `json:"moduleLoader"`
+	ImageRepoSecret *v1.LocalObjectReference `json:"imageRepoSecret,omitempty"`
+	KmomLoaded      bool                     `json:"kmodLoaded"`
+}
+
+// NodeModulesStateSpec describes the desired state of modules on the node
+// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+// +kubebuilder:validation:Required
+type NodeModulesStateSpec struct {
+	// ModuleStateSpec list the spec of all the modules that need to be executed
+	// on the node
+	Modules []ModuleNodeSpec `json:"modules,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+type ModuleNodeStatus struct {
```

Comment:
I think we need a `ModuleLoaderSpec` as well here, to determine if e.g. the desired parameters are different from those the module was loaded with.

Reply:
I am not sure if we want to support such a flow. Changing the parameters requires module unloading and reloading. It seems more like an upgrade flow